app/controllers/PersonalDetailsController.scala

```diff
@@ -31,58 +35,87 @@ object PersonalDetailsController extends PersonalDetailsController(ApplicationCl
-            phone = None
-          ))
-          Ok(views.html.application.generalDetails(formFromUser))
-      }
+        Ok(views.html.application.generalDetails(formFromUser, continueToTheNextStep))
+    }
   }
 
-  def submitGeneralDetails = CSRSecureAppAction(PersonalDetailsRole) { implicit request =>
+  def submitGeneralDetailsAndContinue() = CSRSecureAppAction(EditPersonalDetailsAndContinueRole) { implicit request =>
     implicit user =>
-      implicit val now: LocalDate = LocalDate.now
-      GeneralDetailsForm.form.bindFromRequest.fold(
-        errorForm => {
-          Future.successful(Ok(views.html.application.generalDetails(errorForm)))
-        },
-        gd => {
-          (for {
-            _ <- applicationClient.updateGeneralDetails(user.application.applicationId, user.user.userID,
-              removePostCodeWhenOutsideUK(gd), user.user.email)
-            _ <- userManagementClient.updateDetails(user.user.userID, gd.firstName, gd.lastName, Some(gd.preferredName))
-            redirect <- updateProgress(data => data.copy(user = user.user.copy(firstName = gd.firstName, lastName = gd.lastName,
-              preferredName = Some(gd.preferredName)), application = data.application.map(_.copy(applicationStatus = IN_PROGRESS))
-            ))(_ => Redirect(routes.SchemePreferencesController.present()))
-          } yield {
-            redirect
-          })
-        }
-      )
+      submit(ContinueToNextStepInJourney, Redirect(routes.SchemePreferencesController.present()))
+  }
+
+  def submitGeneralDetails() = CSRSecureAppAction(EditPersonalDetailsRole) { implicit request =>
+    implicit user =>
+      submit(RedirectToTheDashboard, Redirect(routes.HomeController.present()).flashing(success("personalDetails.updated")))
+  }
+
+  private def continuetoTheNextStep(onSuccess: OnSuccess) = onSuccess match {
+    case ContinueToNextStepInJourney => true
+    case RedirectToTheDashboard => false
+  }
+
+  private def submit(onSuccess: OnSuccess, redirectOnSuccess: Result)(
+    implicit user: CachedDataWithApp, hc: HeaderCarrier, request: Request[_]) = {
+    implicit val now: LocalDate = LocalDate.now
+
+    GeneralDetailsForm.form.bindFromRequest.fold(
+      errorForm => Future.successful(Ok(views.html.application.generalDetails(errorForm, continuetoTheNextStep(onSuccess)))),
+      gd => for {
+        _ <- applicationClient.updateGeneralDetails(user.application.applicationId, user.user.userID,
+          removePostCodeWhenOutsideUK(gd), user.user.email, continuetoTheNextStep(onSuccess))
+        _ <- userManagementClient.updateDetails(user.user.userID, gd.firstName, gd.lastName, Some(gd.preferredName))
+        redirect <- updateProgress(data => data.copy(user = user.user.copy(firstName = gd.firstName, lastName = gd.lastName,
+          preferredName = Some(gd.preferredName)), application =
+          if (continuetoTheNextStep(onSuccess)) data.application.map(_.copy(applicationStatus = IN_PROGRESS)) else data.application
```

Comment:
All ok. The use of a case class (OnSuccess) here is interesting and we should have a chat about it at some point.